haystack/generator/transformers.py

```diff
@@ -18,91 +19,227 @@ class RAGGeneratorType(Enum):
         self.max_length = max_length
         self.min_length = min_length
-        self.tokenizer = RagTokenizer.from_pretrained(model_name_or_path)
         self.generator_type = generator_type
+        self.num_beams = num_beams
+        self.embed_title = embed_title
+        self.prefix = prefix
+
+        if use_gpu and torch.cuda.is_available():
+            self.device = torch.device("cuda")
+        else:
+            self.device = torch.device("cpu")
+
+        self.tokenizer = RagTokenizer.from_pretrained(model_name_or_path)
 
         if self.generator_type == RAGGeneratorType.SEQUENCE:
             self.model = RagSequenceForGeneration.from_pretrained(model_name_or_path)
         else:
             self.model = RagTokenForGeneration.from_pretrained(model_name_or_path)
 
+        self.passage_tokenizer = DPRContextEncoderTokenizerFast.from_pretrained(passage_embedding_model)
+        self.passage_encoder = DPRContextEncoder.from_pretrained(passage_embedding_model).to(self.device)
+
+    # Copied cat_input_and_doc method from transformers.RagRetriever
+    def _cat_input_and_doc(self, doc_title: str, doc_text: str, input_string: str, prefix: Optional[str]):
+        if doc_title.startswith('"'):
+            doc_title = doc_title[1:]
+        if doc_title.endswith('"'):
+            doc_title = doc_title[:-1]
+        if prefix is None:
+            prefix = ""
+        out = (prefix + doc_title + self.model.config.title_sep + doc_text + self.model.config.doc_sep +
+               input_string).replace("  ", " ")
+
+        return out
 
-    # TODO: Copy from RagRetriever and modified
-    # Need to understand the logic
-    def _postprocess_docs(self, retrieved_texts: List[str], question: str, return_tensors: str = "pt"):
-        def cat_input_and_doc(doc_text, input_string):
-            out = (" / " + doc_text + " // " + input_string).replace("  ", " ")
-            return out
+    # Copied postprocess_docs method from transformers.RagRetriever and modified
+    def _get_contextualized_embedding(self, texts: List[str], question: str, titles: Optional[List[str]] = None,
+                                      return_tensors: str = "pt"):
+
+        titles_list = titles if self.embed_title and titles is not None else [""] * len(texts)
+        prefix = self.prefix if self.prefix is not None else self.model.config.generator.prefix
 
         rag_input_strings = [
-            cat_input_and_doc(
-                retrieved_texts[i],
-                question,
+            self._cat_input_and_doc(
+                doc_title=titles_list[i],
+                doc_text=texts[i],
+                input_string=question,
+                prefix=prefix
             )
-            for i in range(len(retrieved_texts))
+            for i in range(len(texts))
         ]
 
-        contextualized_inputs = self.tokenizer.question_encoder.batch_encode_plus(
+        contextualized_embedding = self.tokenizer.question_encoder.batch_encode_plus(
```

Comment:
Naming of `contextualized_inputs`was better before. This is still just the input for the model (raw text => tokens => input_ids), while the embedding will be produced later in the model.
Similarly, the method name should be switched `_get_contextualized_embedding` => `_get_contextualized_inputs`